service/controller/clusterapi/v29/resource/tccp/template_render_test.go

```diff
@@ -0,0 +1,273 @@
+package tccp
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"flag"
+	"io/ioutil"
+	"path/filepath"
+	"strconv"
+	"testing"
+	"unicode"
+
+	g8sv1alpha1 "github.com/giantswarm/apiextensions/pkg/apis/cluster/v1alpha1"
+	"github.com/giantswarm/micrologger/microloggertest"
+	"github.com/google/go-cmp/cmp"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+
+	"github.com/giantswarm/aws-operator/pkg/annotation"
+	"github.com/giantswarm/aws-operator/pkg/label"
+	"github.com/giantswarm/aws-operator/service/controller/clusterapi/v29/controllercontext"
+	"github.com/giantswarm/aws-operator/service/controller/clusterapi/v29/detection"
+	"github.com/giantswarm/aws-operator/service/controller/clusterapi/v29/key"
+)
+
+const (
+	TestClusterID           = "8y5ck"
+	TestMachineDeploymentID = "al9qy"
+)
+
+var update = flag.Bool("update", false, "update .golden CF template file")
+
+// Test_newTemplateBody tests tenant cluster CloudFormation template rendering.
+// It is meant to be used as a tool to easily check resulting CF template and
+// prevent from accidental CF template changes.
+//
+// It uses golden file as reference template and when changes to template are
+// intentional, they can be updated by providing -update flag for go test.
+//
+//  go test ./service/controller/clusterapi/v29/resource/tccp -run Test_newTemplateBody -update
+//
+func Test_newTemplateBody(t *testing.T) {
+	testCases := []struct {
+		name         string
+		cr           v1alpha1.Cluster
+		ctx          controllercontext.Context
+		tp           templateParams
+		errorMatcher func(error) bool
+	}{
+		{
+			name: "case 0: basic test",
+			cr:   defaultCluster(),
+			ctx:  defaultControllerContext(),
+			tp: templateParams{
+				DockerVolumeResourceName:   "rsc-abbacd01",
+				MasterInstanceResourceName: "rsc-ac0dc01",
+			},
+			errorMatcher: nil,
+		},
+	}
+
+	var r *Resource
+	{
+		d, err := detection.New(detection.Config{Logger: microloggertest.New()})
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		c := Config{
+			Detection:        d,
+			Logger:           microloggertest.New(),
+			EncrypterBackend: "kms",
+			VPCPeerID:        "vpc-f8d0e10b",
+		}
+
+		r, err = New(c)
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	for i, tc := range testCases {
+		t.Run(strconv.Itoa(i), func(t *testing.T) {
+			ctx := controllercontext.NewContext(context.Background(), tc.ctx)
+			actual, err := r.newTemplateBody(ctx, tc.cr, tc.tp)
+
+			switch {
+			case err == nil && tc.errorMatcher == nil:
+				// correct; carry on
+			case err != nil && tc.errorMatcher == nil:
+				t.Fatalf("error == %#v, want nil", err)
+			case err == nil && tc.errorMatcher != nil:
+				t.Fatalf("error == nil, want non-nil")
+			case !tc.errorMatcher(err):
+				t.Fatalf("error == %#v, want matching", err)
+			}
+
+			golden := filepath.Join("testdata", normalizeToFileName(tc.name)+".golden")
```

Comment:
Why not use `.yaml` extension here?

Reply:
General practise between projects "out there" uses `.golden` for these files. I'd also personally lean towards that because it emphasizes the purpose of that file clearly.

Reply:
When they are in `testdata` directory the purpose is 100% clean to me and we lose some editor feature which may ease editing those (at least my editor isn't smart enough).

Reply:
My other argument is I would also like to copy this into devctl for generating files functionality. Then extensions would be even nicer so we can check formatting/highlighting etc of generated (not only) Go files.

Reply:
E.g. in Vim I'd just use `:set filetype` but for me the file extension here is not the important bit but the overall concept and approach to implement it so I changed it to `.yaml`.

Reply:
I educated myself more from the links you provided (sorry missed them at first). And since this is community standard I'd be keen to revert them to `.golden` (I can do it myself) 🙈

Reply:
Sure, either way. I like `.golden` :)